assemblyscript/assembly/convert_image/host/as-wasi-nn.ts

```diff
@@ -0,0 +1,241 @@
+/**
+ * Export the wasi-nn APIs with an ergonomic, AssemblyScript-compatible binding. We expect users to
+ * use these bindings, not the raw `wasi_ephemeral_nn.ts` bindings.
+ *
+ * The equivalent Rust bindings (i.e. high-level, ergonomic) for wasi-nn currently live as an
+ * example in the Wasmtime repository--they should be moved to their own repository at some point.
+ * For comparison, those Rust bindings look like in part (see
+ * https://github.com/bytecodealliance/wasmtime/blob/main/crates/wasi-nn/examples/wasi-nn-rust-bindings/src/generated.rs#L57-L61):
+ *
+ *   pub unsafe fn load(builder: GraphBuilderArray, encoding: GraphEncoding, target:
+ *     ExecutionTarget) -> Result<Graph>
+ *
+ * Our AssemblyScript bindings should expose this level of functionality until interface types or
+ * WITX tooling make this project irrelevant.
+ */
+import * as wasi_ephemeral_nn from './wasi_ephemeral_nn';
+
+/**
+ * A machine learning model.
+ */
+export class Graph {
+    private constructor(private pointer: i32) { }
+
+    /**
+     * Create a `Graph` from one or more binary blobs.
+     * @param builder the binary blobs that make up the graph
+     * @param encoding the framework required for
+     * @param target the device on which to run the graph
+     * @returns an initialized `Graph`
+     */
+    static load(builder: u8[][], encoding: GraphEncoding, target: ExecutionTarget): Graph {
+        let graphBuilder: u32[] = [];
+
+        for (let i = 0; i < builder.length; i++) {
+            graphBuilder.push(getArrayPtr(builder[i]));
+            graphBuilder.push(builder[i].length);
+        }
+
+        let graphPointer: u32 = changetype<u32>(memory.data(4));
```

Comment:
```suggestion
        let graphPointer = memory.data(4) as u32;
```

Reply:
Why is `as u32` preferred to using changetype in this situation?

Reply:
`changetype` is special cast which usually uses for unsafe cast some reference to `usize` (or u32) and wise versa. `memory.data` return `usize` which is `u32` in wasm32 and `u64` on wasm64 (doesn't support yet by spec). For int <-> int cast you should use ordinal type casting via `<T>val` or `val as T`

Reply:
Thanks for the info!